feat: add svelte compiler

userquin opened the pull request:
This PR includes (don't merge yet):
- add `esno` as dev dependency
- add `IntelliJ/Webstorm` stuff on `.gitignore`
- remove single quotes on root `package.json` scripts to allow build and dev on windows
- `svelte` compiler
- include some `svelte` entries on `README.md`
- `vite + svelte-ts` example
- `vite + svelte-kit` example: this example is not working, we need to fix it: created with `pnpm init svelte@next vite-svelte-kit`:
```shell
√ Which Svelte app template? » Skeleton project
√ Use TypeScript? ... No / **Yes**
√ Add ESLint for code linting? ... No / **Yes**
√ Add Prettier for code formatting? ... **No** / Yes
✔ Copied project files
✔ Added TypeScript support. To use it inside Svelte components, add lang="ts" to the attributes of a script tag.
✔ Added ESLint.
Readme for ESLint and Svelte: https://github.com/sveltejs/eslint-plugin-svelte3
```

We need to review how to augment `svetle` types (if can be done), not yet included the file `types/svelte.d.ts`.

bluwy commented on src/core/compilers/svelte.ts:
```diff
@@ -0,0 +1,18 @@
+    filename: `${collection}-${icon}.svelte`,
+    format: options.format || 'esm',
+    legacy: typeof options.legacy !== 'undefined' ? options.legacy : false,
+    generate: options.generate || 'dom',
```
We might need to handle this automatically for ssr, using the [ssr flag](https://vitejs.dev/guide/ssr.html#ssr-specific-plugin-logic) from vite. Though this is an unplugin and I'm not sure if it's possible here.

I think it's best if we offload this to the bundler's svelte plugin though, similar to solid's integration, so options are also inherited when configuring the svelte plugin.

userquin replied:
we can use `import.meta.env.SSR` or `process.env.SSR` and remove it from options

bluwy replied:
I don't think Vite supports that, or maybe it's unplugin specific?

userquin replied:
> 
> 
> I don't think Vite supports that, or maybe it's unplugin specific?

this plugin has been changed to delegate the compilation to `vite-plugin-svelte`, so we shouldn't have problems with SSR, the compiled component looks good, see last comment

userquin commented:
I split `svelte` into 2 compilers, former (called `svelte`) to be used with `vite` and later to be used with `svelte-kit` (called `svelte-kit`. 

Since `svelte-kit` will include `vite-plugin-svelte` plugin, I just add `enforce: 'pre'` and add the `.svelte` extension on `resolveId`  hook and let the `svelte-plugin` do its work: the compiler just returns the `svg` text.

@bluwy Is there a way to remove this warnings when `vite-plugin-svelte` resolves the virtual request for icons?
<details><summary><strong>vite-plugin-svelte warnings</strong></summary>
<p>
<pre>
Sourcemap for "~icons/logos/svelte-icon.svelte" points to missing source files
Sourcemap for "~icons/mdi/store-24-hour.svelte" points to missing source files
Sourcemap for "~icons/mdi/alarm-off.svelte" points to missing source files
Sourcemap for "~icons/icon-park/abnormal.svelte" points to missing source files
</pre>
</details>

@antfu I've updated the `unplugin` to latest version `0.2.8`, when running `svelte-kit` with `"type": "module"`, to solve the error (isomorphic `__dirname` from your blog):

<details><summary><strong>__dirname error</strong></summary>
<p>
<pre>
__dirname is not defined
ReferenceError: __dirname is not defined
    at file:///F:/work/projects/quini/GitHub/antfu/unplugin-icons/node_modules/.pnpm/unplugin@0.2.3_rollup@2.56.3+vite@2.5.1/node_modules/unplugin/dist/index.mjs:226:40
    at ModuleJob.run (node:internal/modules/esm/module_job:154:23)
    at async Loader.import (node:internal/modules/esm/loader:166:24)
    at async load_config (file:///F:/work/projects/quini/GitHub/antfu/unplugin-icons/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.165_svelte@3.42.5/node_modules/@sveltejs/kit/dist/cli.js:701:17)
    at async get_config (file:///F:/work/projects/quini/GitHub/antfu/unplugin-icons/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.165_svelte@3.42.5/node_modules/@sveltejs/kit/dist/cli.js:773:10)
    at async file:///F:/work/projects/quini/GitHub/antfu/unplugin-icons/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.165_svelte@3.42.5/node_modules/@sveltejs/kit/dist/cli.js:833:18
</pre>
</details>

I cannot run the `vite-vue3` example project (the `Vue2` example just works), seems there is something wrong resolving the `vue` types: I just go to the example folder and run `pnpm run dev`.

<details><summary><strong>Vue3 browser error</strong></summary>
<p>
<pre style="color: red;">
Uncaught SyntaxError: The requested module '/node_modules/.vite/vue.js?v=e2c6055a' does not provide an export named 'createElementBlock'
</pre>
</p>
</details>

I cannot run also `React` example:
<details><summary><strong>React browser errors</strong></summary>
<p>
<pre style="color: red;">
react.jsx:4 Uncaught ReferenceError: React is not defined
    at logosReact (react.jsx:4)
    at renderWithHooks (react-dom.development.js:14985)
    at mountIndeterminateComponent (react-dom.development.js:17811)
    at beginWork (react-dom.development.js:19049)
    at HTMLUnknownElement.callCallback2 (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at beginWork$1 (react-dom.development.js:23964)
    at performUnitOfWork (react-dom.development.js:22776)
    at workLoopSync (react-dom.development.js:22707)
logosReact @ react.jsx:4
renderWithHooks @ react-dom.development.js:14985
mountIndeterminateComponent @ react-dom.development.js:17811
beginWork @ react-dom.development.js:19049
callCallback2 @ react-dom.development.js:3945
invokeGuardedCallbackDev @ react-dom.development.js:3994
invokeGuardedCallback @ react-dom.development.js:4056
beginWork$1 @ react-dom.development.js:23964
performUnitOfWork @ react-dom.development.js:22776
workLoopSync @ react-dom.development.js:22707
renderRootSync @ react-dom.development.js:22670
performSyncWorkOnRoot @ react-dom.development.js:22293
scheduleUpdateOnFiber @ react-dom.development.js:21881
updateContainer @ react-dom.development.js:25482
(anonymous) @ react-dom.development.js:26021
unbatchedUpdates @ react-dom.development.js:22431
legacyRenderSubtreeIntoContainer @ react-dom.development.js:26020
render @ react-dom.development.js:26103
(anonymous) @ main.tsx:6
react-dom.development.js:20085 The above error occurred in the <logosReact> component:

    at logosReact (http://localhost:3000/@id/~icons/logos/react.jsx:2:3)
    at header
    at div
    at App (http://localhost:3000/src/App.tsx:31:29)

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://reactjs.org/link/error-boundaries to learn more about error boundaries.
logCapturedError @ react-dom.development.js:20085
update.callback @ react-dom.development.js:20118
callCallback @ react-dom.development.js:12318
commitUpdateQueue @ react-dom.development.js:12339
commitLifeCycles @ react-dom.development.js:20736
commitLayoutEffects @ react-dom.development.js:23426
callCallback2 @ react-dom.development.js:3945
invokeGuardedCallbackDev @ react-dom.development.js:3994
invokeGuardedCallback @ react-dom.development.js:4056
commitRootImpl @ react-dom.development.js:23151
unstable_runWithPriority @ scheduler.development.js:468
runWithPriority$1 @ react-dom.development.js:11276
commitRoot @ react-dom.development.js:22990
performSyncWorkOnRoot @ react-dom.development.js:22329
scheduleUpdateOnFiber @ react-dom.development.js:21881
updateContainer @ react-dom.development.js:25482
(anonymous) @ react-dom.development.js:26021
unbatchedUpdates @ react-dom.development.js:22431
legacyRenderSubtreeIntoContainer @ react-dom.development.js:26020
render @ react-dom.development.js:26103
(anonymous) @ main.tsx:6
react.jsx:4 Uncaught ReferenceError: React is not defined
    at logosReact (react.jsx:4)
    at renderWithHooks (react-dom.development.js:14985)
    at mountIndeterminateComponent (react-dom.development.js:17811)
    at beginWork (react-dom.development.js:19049)
    at HTMLUnknownElement.callCallback2 (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at beginWork$1 (react-dom.development.js:23964)
    at performUnitOfWork (react-dom.development.js:22776)
    at workLoopSync (react-dom.development.js:22707)
</pre>
</details>

userquin commented:
Upps, seems deps have been updated...

antfu commented:
Can't we have a single compiler for both or them? What's the problem / difficulty?

> Vue 3 example

Yeah, I knew. Just don't know how to solve it yet as pnpm is somehow failed to differciate a package in two versions somehow.

userquin commented:
just take a look, svelte + vite will use `svelte.compile` while for `svelte-kit` we use its plugin. Do you mean that shoould work with both? I'll test it.

userquin commented:
@antfu, the `vue3` is broken:
```shell
> @ dev F:\work\projects\quini\GitHub\antfu\unplugin-icons\examples\vite-vue3
> cross-env DEBUG=unplugin-icons:* vite

failed to load config from F:\work\projects\quini\GitHub\antfu\unplugin-icons\examples\vite-vue3\vite.config.ts
error when starting dev server:
TypeError: Cannot read property 'prefix' of undefined
    at ViteComponentsResolver (F:\work\projects\quini\GitHub\antfu\unplugin-icons\dist\resolver.js:15:13)
    at Object.<anonymous> (F:\work\projects\quini\GitHub\antfu\unplugin-icons\examples\vite-vue3\vite.config.ts:39:37)
    at Module._compile (node:internal/modules/cjs/loader:1092:14)
    at Object.require.extensions.<computed> [as .ts] (F:\work\projects\quini\GitHub\antfu\unplugin-icons\node_modules\.pnpm\vite@2.5.1\node_modules\vite\dist\node\chunks\dep-972722fa.js:75935:20)
    at Module.load (node:internal/modules/cjs/loader:972:32)
    at Function.Module._load (node:internal/modules/cjs/loader:813:14)
    at Module.require (node:internal/modules/cjs/loader:996:19)
    at require (node:internal/modules/cjs/helpers:92:18)
    at loadConfigFromBundledFile (F:\work\projects\quini\GitHub\antfu\unplugin-icons\node_modules\.pnpm\vite@2.5.1\node_modules\vite\dist\node\chunks\dep-972722fa.js:75943:17)
    at loadConfigFromFile (F:\work\projects\quini\GitHub\antfu\unplugin-icons\node_modules\.pnpm\vite@2.5.1\node_modules\vite\dist\node\chunks\dep-972722fa.js:75862:32)
 ERROR  Command failed with exit code 1.
```

userquin commented:
```ts
 IconsResolver({
  prefix: '',
}),
```

userquin commented:
> 
> 
> Can't we have a single compiler for both or them? What's the problem / difficulty?

only one compiler just works man

userquin commented:
@bluwy I've update the compiler to delegate the compilation on vite and svelte kit to `vite-plugin-svelte`.

userquin commented:
@antfu I forgot to mention that I change the `loader.ts` module to load `@iconify/json-tools` with:
```ts
import pkg from '@iconify/json-tools'

const { SVG, Collection } = pkg
```
instead using`import` directly: the module will fail when used on `esm` since there is no `mjs` version:
```shell
F:\work\projects\quini\GitHub\antfu\unplugin-icons\examples\vite-svelte-kit>pnpm run dev

> @ dev F:\work\projects\quini\GitHub\antfu\unplugin-icons\examples\sveltekit
> svelte-kit dev

Malformed svelte.config.js
file:///F:/work/projects/quini/GitHub/antfu/unplugin-icons/dist/chunk-L5UNGFYX.mjs:6
import { SVG, Collection } from "@iconify/json-tools";
         ^^^
SyntaxError: Named export 'SVG' not found. The requested module '@iconify/json-tools' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from '@iconify/json-tools';
const { SVG, Collection } = pkg;

    at ModuleJob._instantiate (node:internal/modules/esm/module_job:105:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:151:5)
    at async Loader.import (node:internal/modules/esm/loader:166:24)
    at async load_config (file:///F:/work/projects/quini/GitHub/antfu/unplugin-icons/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.165_svelte@3.42.5/node_modules/@sveltejs/kit/dist/cli.js:701:17)
    at async get_config (file:///F:/work/projects/quini/GitHub/antfu/unplugin-icons/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.165_svelte@3.42.5/node_modules/@sveltejs/kit/dist/cli.js:773:10)
    at async file:///F:/work/projects/quini/GitHub/antfu/unplugin-icons/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.165_svelte@3.42.5/node_modules/@sveltejs/kit/dist/cli.js:833:18
```

bluwy reviewed (commented):
Do we need to add the svelte compiler [here](https://github.com/antfu/unplugin-icons/blob/main/src/core/loader.ts)? Maybe an example of using `onwarn` in the examples to suppress the sourcemap warning would be helpful too

userquin commented:
> 
> 
> Do we need to add the svelte compiler [here](https://github.com/antfu/unplugin-icons/blob/main/src/core/loader.ts)? Maybe an example of using `onwarn` in the examples to suppress the sourcemap warning would be helpful too

It is required, but can be removed and just return the svgText: I can add the onwarn on vite-svelte and sveltekit examples if you want

userquin commented:
@bluwy the compiler is **NOT** your `svelte compiler`, it is the `internal unplugin-icons compiler`, we are not compiling anymore svelte code, take a look at https://github.com/antfu/unplugin-icons/pull/47/files#diff-9604cc434156d08714f75e10b42234d1f64535b91c86169e2f9d2af0db56207d

bluwy commented:
Yeah I know that. I was reviewing the code in the GitHub mobile app and for some reason it didn't show the diff for that file. So lgtm 👍 The onwarn example would be a nice to have.

userquin commented:
> 
> 
> Yeah I know that. I was reviewing the code in the GitHub mobile app and for some reason it didn't show the diff for that file. So lgtm 👍 The onwarn example would be a nice to have.

Where should be configure, on root options? Seems on dev not working, and on build + preview the warnings not logged: I've found this entry, you comment about it https://giters.com/sveltejs/kit/issues/2253

bluwy commented:
Looks like I was mistaken, the warning came from Vite [here](https://github.com/vitejs/vite/blob/0693d038e096dd4db3140507269c6aaeefd2186e/packages/vite/src/node/server/sourcemap.ts#L51). It's trying to load the virtual path but failed. I think it should load via `pluginContainer.load` though, but I'm not too familiar with sourcemaps to know if that's correct. Maybe best to leave it as-is for now since it could be an issue upstream.

dominikg commented:
examples/vite-svelte/svelte.config.ts should be svelte.config.js,  svelte does not support ts config

userquin commented:
> 
> 
> examples/vite-svelte/svelte.config.ts should be svelte.config.js, svelte does not support ts config

This project is a vite project and so vite will transform it, the project just works.

dominikg commented:
well, neither vite-plugin-svelte nor svelte vscode extension or any other tool in the svelte ecosystem reads svelte.config.ts

even if it works here if you import and pass it through inline in vite config, i'd consider this an anti-pattern, especially in an example

userquin commented:
> well, neither vite-plugin-svelte nor svelte vscode extension or any other tool in the svelte ecosystem reads svelte.config.ts
> 
> even if it works here if you import and pass it through inline in vite config, i'd consider this an anti-pattern, especially in an example

Ok. I Will change it, maybe tmr...

antfu commented:
I have refactored a bit for the project structure, please rebase and apply the new style. Thanks.

userquin commented:
> 
> 
> I have refactored a bit for the project structure, please rebase and apply the new style. Thanks.

I'm having problems to rebase, some cherry-pick commit, there is no way man: I will add a new PR since the changes are minimal...

userquin commented:
superseded by #49
